### PR TITLE
Add `reachability-metadata.json` to all Log4j modules

### PR DESCRIPTION
**Pull request by ppkarwasz**
This PR adds a plugin processor that generates a GraalVM reachability metadata file and bundles it with each Log4j module.

After this change, GraalVM will be able to handle Log4j Core without any user intervention.

**Note**: I also added a static `resource-config.json` file that specifies that all resources starting with `log4j2` and the plugin descriptor must be included in the GraalVM image.

Part of #2831.

### CODE REVIEW

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/internal/ReachabilityMetadata.java [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+
+        public void writeArrayEnd() throws IOException {
+            output.append(']');
+        }
+    }
+
+    /**
+     * Specifies a field that needs to be accessed through reflection.
+     */
+    public static class Field implements Comparable<Field> {
```
Comment (vy) [inline log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/internal/ReachabilityMetadata.java:137]: This and other classes can be `final` too.

File: log4j-core/src/main/resources/META-INF/native-image/org.apache.logging.log4j/log4j-core/resource-config.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "resources": {
+    "includes": [
+      {
+        "pattern": "log4j2.*"
```
Comment (vy) [inline log4j-core/src/main/resources/META-INF/native-image/org.apache.logging.log4j/log4j-core/resource-config.json:5]: What is the purpose of this particular include entry? (I get the `META-INF/.../Log4j2Plugins.dat` one.)

File: log4j-parent/pom.xml [hunk 3]
```diff
@@ -576,6 +576,12 @@
         <version>${jna.version}</version>
       </dependency>
 
+      <dependency>
+        <groupId>net.javacrumbs.json-unit</groupId>
+        <artifactId>json-unit-assertj</artifactId>
+        <version>${json-unit.version}</version>
+      </dependency>
+
```
Comment (vy) [inline log4j-parent/pom.xml:584]: Both `json-unit` and `json-unit-assertj` are only used in `log4j-core-test`, shall we move them there instead?

**Review by vy — CHANGES_REQUESTED** [review]
Slick implementation! :100:

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessorTest.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                .getClassLoader()
+                .getResources("META-INF/native-image/org.apache.logging.log4j/log4j-core-test/reflect-config.json"))) {
+            if (url.getPath().contains("test-classes")) {
+                reachabilityMetadataUrl = url;
+                break;
+            }
+        }
+        Assertions.assertThat(reachabilityMetadataUrl).isNotNull();
+        reachabilityMetadata =
+                IOUtils.toString(Objects.requireNonNull(reachabilityMetadataUrl), StandardCharsets.UTF_8);
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessorTest.java]: Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/0717df689258756b6e4a9da1225e6c1cd3bb3470

I thought IDEA will not understand the assertions and keep complaining about the nullability. :wink:

File: log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessorTest.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+import org.apache.commons.io.IOUtils;
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class GraalVmProcessorTest {
+
+    private static final String FAKE_PLUGIN = String.join(
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core-test/src/test/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessorTest.java]: Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/d58894f84f4b00e491a9b77f4b30deb935fdcc98

Actually JsonUnit apparently understands maps (it delegates them to whatever JSON mapper is on the classpath).

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+            // Many users will have `log4j-core` on the annotation processor path, but do not have Log4j Plugins.
+            // Therefore, we check for the annotation processor required options only if some elements were processed.
+            //
+            String groupId = processingEnv.getOptions().get("log4j.graalvm.groupId");
+            String artifactId = processingEnv.getOptions().get("log4j.graalvm.artifactId");
+            if (groupId == null || artifactId == null) {
+                messager.printMessage(
+                        Diagnostic.Kind.ERROR,
+                        "The `" + GraalVmProcessor.class.getName()
+                                + "` annotation processor is missing the required `maven.groupId` and `maven.artifactId` options.\n"
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java]: Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/0717df689258756b6e4a9da1225e6c1cd3bb3470

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+                return false;
+            }
+            String reachabilityMetadataPath =
+                    String.format("META-INF/native-image/%s/%s/reflect-config.json", groupId, artifactId);
+            try {
+                messager.printMessage(
+                        Diagnostic.Kind.NOTE,
+                        String.format(
+                                "%s: writing GraalVM metadata for %d Java classes to `%s`.",
+                                GraalVmProcessor.class.getSimpleName(),
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java]: Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/0717df689258756b6e4a9da1225e6c1cd3bb3470

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+                                reachableTypes.size(),
+                                reachabilityMetadataPath));
+                writeReachabilityMetadata(reachabilityMetadataPath, processedElements.toArray(new Element[0]));
+            } catch (IOException e) {
+                StringWriter sw = new StringWriter();
+                sw.append(GraalVmProcessor.class.getSimpleName())
+                        .append(": unable to write reachability metadata to file ")
+                        .append(reachabilityMetadataPath)
+                        .append("\n");
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java]: Yes, we probably can throw here.

Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/0717df689258756b6e4a9da1225e6c1cd3bb3470.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+            return type.cast(element);
+        }
+        processingEnv
+                .getMessager()
+                .printMessage(
+                        Diagnostic.Kind.ERROR,
+                        "Unexpected type of element `" + element + "`: expecting `" + type.getName() + "` but found `"
+                                + element.getClass().getName() + "`",
+                        element);
+        return null;
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java]: Yes, this case should never happen anyway. Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/0717df689258756b6e4a9da1225e6c1cd3bb3470 together with the null-checks.

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+                    }
+                }
+                break;
+            case PARAMETER:
+                // Do nothing, the containing method must be annotated with a factory annotation.
+                break;
+            default:
+                processingEnv
+                        .getMessager()
+                        .printMessage(Diagnostic.Kind.ERROR, "Invalid Log4j Attribute element.", element);
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/GraalVmProcessor.java]: Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/0717df689258756b6e4a9da1225e6c1cd3bb3470

File: log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/internal/ReachabilityMetadata.java [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+    /**
+     * Key used to specify the list of fields available for reflection.
+     */
+    public static final String FIELDS = "fields";
+    /**
+     * Key used to specify the list of methods available for reflection.
+     */
+    public static final String METHODS = "methods";
+
+    private static class MinimalJsonWriter {
```
[the remark this answers is not in the document]
Comment (ppkarwasz) [reply log4j-core/src/main/java/org/apache/logging/log4j/core/config/plugins/processor/internal/ReachabilityMetadata.java]: Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/0537fae39ec73722f79381ab4a680ac3b2e5a62b

File: log4j-core/src/main/resources/META-INF/native-image/org.apache.logging.log4j/log4j-core/resource-config.json (hunk 2 above)
Comment (ppkarwasz) [reply log4j-core/src/main/resources/META-INF/native-image/org.apache.logging.log4j/log4j-core/resource-config.json:5]: This entry is for configuration files of the form `log4j2<contextName>.<extension>`. (Unfortunately the `resource-config.json` schema leavers no fields for comments).

Thanks for the comment, that reminded me that `log4j-api` needs a configuration file too for `log4j2.component.properties`, `log4j2.simplelog.properties` and `log4j2.StatusLogger.properties`. Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/79909eab2699b9f5032c2465a1341f2b175e01aa

File: log4j-parent/pom.xml (hunk 3 above)
Comment (ppkarwasz) [reply log4j-parent/pom.xml]: Fixed in https://github.com/apache/logging-log4j2/pull/3053/commits/d25e0e6f632fa47f36afaadb7e2af0e4452560bc

**Review by vy — APPROVED** [review]

**Pull request merged**